display_list/effects/dl_image_filter.h

```diff
@@ -741,6 +750,82 @@ class DlLocalMatrixImageFilter final : public DlImageFilter {
   std::shared_ptr<const DlImageFilter> image_filter_;
 };
 
+class DlRuntimeEffectImageFilter final : public DlImageFilter {
+ public:
+  explicit DlRuntimeEffectImageFilter(
+      sk_sp<DlRuntimeEffect> runtime_effect,
+      std::vector<std::shared_ptr<DlColorSource>> samplers,
+      std::shared_ptr<std::vector<uint8_t>> uniform_data)
+      : runtime_effect_(std::move(runtime_effect)),
+        samplers_(std::move(samplers)),
+        uniform_data_(std::move(uniform_data)) {}
+
+  std::shared_ptr<DlImageFilter> shared() const override {
+    return std::make_shared<DlRuntimeEffectImageFilter>(
+        this->runtime_effect_, this->samplers_, this->uniform_data_);
+  }
+
+  static std::shared_ptr<DlImageFilter> Make(
+      sk_sp<DlRuntimeEffect> runtime_effect,
+      std::vector<std::shared_ptr<DlColorSource>> samplers,
+      std::shared_ptr<std::vector<uint8_t>> uniform_data) {
+    return std::make_shared<DlRuntimeEffectImageFilter>(
+        std::move(runtime_effect), std::move(samplers),
+        std::move(uniform_data));
+  }
+
+  DlImageFilterType type() const override {
+    return DlImageFilterType::kRuntimeEffect;
+  }
+  size_t size() const override { return sizeof(*this); }
+
+  bool modifies_transparent_black() const override { return false; }
+
+  SkRect* map_local_bounds(const SkRect& input_bounds,
+                           SkRect& output_bounds) const override {
+    output_bounds = input_bounds;
+    return &output_bounds;
```

Comment:
This gets tricky. Can they implement a custom image filter that does something like a blur - i.e. expands the bounds of the pixels?

Reply:
We talked about this offline and discussed that there doesn't seem to be a good way to express all of the general purpose transformations that can be done with an image filter.